src/DataSet/ObjectDataSet.php

```diff
@@ -13,39 +14,113 @@
 use Yiisoft\Validator\RulesProviderInterface;
 
 use function array_key_exists;
+use function get_class;
 
 /**
  * This data set makes use of attributes introduced in PHP 8. It simplifies rules configuration process, especially for
- * nested data and relations. Please refer to the guide for example.
+ * nested data and relations. Please refer to the guide for examples.
  *
  * @link https://www.php.net/manual/en/language.attributes.overview.php
  */
 final class ObjectDataSet implements RulesProviderInterface, DataSetInterface
 {
     private bool $dataSetProvided;
-
     /**
      * @var ReflectionProperty[]
      */
-    private array $reflectionProperties = [];
-
-    private iterable $rules;
+    private ?array $reflectionProperties = null;
+    private ?array $data = null;
+
+    #[ArrayShape([
+        [
+            'rules' => 'iterable',
+            'propertyVisibility' => 'int',
+        ],
+    ])]
+    private static array $cache = [];
+    private string $cacheKey;
 
     public function __construct(
         private object $object,
-        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE|ReflectionProperty::IS_PROTECTED|ReflectionProperty::IS_PUBLIC
+        private int $propertyVisibility = ReflectionProperty::IS_PRIVATE |
+        ReflectionProperty::IS_PROTECTED |
+        ReflectionProperty::IS_PUBLIC
     ) {
-        $this->parseObject();
+        $this->dataSetProvided = $this->object instanceof DataSetInterface;
+        $this->cacheKey = get_class($this->object);
```

Comment:
Cache key must take into account class name and property visibility.

Reply:
OK, but unfortunately `benchGetRulesWithNewInstanceAndNoCache()` no longer works correctly. Can be removed for now I guess, since profit of cache is already proven.